File: library-core/src/main/java/no/wtw/android/restserviceutils/exceptions/RestServiceException.java
package no.wtw.android.restserviceutils.exceptions;

import org.springframework.http.HttpStatus;
import org.springframework.http.converter.HttpMessageConversionException;
import org.springframework.web.client.ResourceAccessException;
import org.springframework.web.client.RestClientException;

import java.io.IOException;

import no.wtw.android.restserviceutils.RestServiceErrorObject;

public class RestServiceException extends IOException {

    private HttpStatus statusCode;
    private RestServiceErrorObject errorObject;

    public RestServiceException(HttpStatus statusCode, String message, RestServiceErrorObject errorObject) {
        this(message);
        setStatusCode(statusCode);
        this.errorObject = errorObject;
        System.out.println(toString());
    }

    public RestServiceException(HttpStatus statusCode, String message, Throwable cause) {
        this(message);
        initCause(cause);
        setStatusCode(statusCode);
        System.out.println(toString());
    }

    public RestServiceException(HttpStatus statusCode, String message) {
        this(message);
        setStatusCode(statusCode);
        System.out.println(toString());
    }

    public RestServiceException(String message) {
        super(message == null || message.equals("") ? "Unknown error" : message);
    }

    private void setStatusCode(HttpStatus statusCode) {
        this.statusCode = statusCode == null ? HttpStatus.I_AM_A_TEAPOT : statusCode;
    }

    public static RestServiceException getInstance(Exception e) {
        Exception cause = (Exception) e.getCause();
        if (cause != null && cause instanceof RestServiceException)
            return (RestServiceException) cause;
        if (e instanceof RestServiceException)
            return (RestServiceException) e;

        if (e instanceof HttpMessageConversionException)
            return new RestServiceException(HttpStatus.INTERNAL_SERVER_ERROR, "Could not parse content", e);

        if (e instanceof ResourceAccessException)
            return new RestServiceException(HttpStatus.INTERNAL_SERVER_ERROR, "Could not access remote server", e);

        if (e instanceof RestClientException)
            return new RestServiceException(HttpStatus.INTERNAL_SERVER_ERROR, "Unknown client error", e);

        return new RestServiceException(HttpStatus.INTERNAL_SERVER_ERROR, e.getMessage(), e);
    }

    public HttpStatus getStatusCode() {
        return statusCode;
    }

    public RestServiceErrorObject getErrorObject() {
        return errorObject;
    }

    @Override
    public String getMessage() {
        RestServiceErrorObject eo = getErrorObject();
        if (eo != null) {
            String message = eo.getMessage();
            if (!message.equals(""))
                return message;
        }
        return super.getMessage();
    }

    @Override
    public String toString() {
        String toString = statusCode.toString() + " " + statusCode.name() + " - " + super.getMessage();
        if (errorObject != null) {
            String errorObjectMessage = errorObject.getMessage();
            if (errorObjectMessage != null && !errorObjectMessage.equals(""))
                toString += " (" + errorObjectMessage + ")";
        }
        if (getCause() != null)
            toString += "\nCaused by: " + getCause().toString() + "";
        return toString;
    }
}

File: library-core/src/main/java/no/wtw/android/restserviceutils/exceptions/RestServiceException.kt
package no.wtw.android.restserviceutils.exceptions

import no.wtw.android.restserviceutils.GsonSingleton
import no.wtw.android.restserviceutils.HttpStatusCode
import no.wtw.android.restserviceutils.RestServiceErrorObject
import okhttp3.Response
import java.io.IOException

class RestServiceException(
    val statusCode: HttpStatusCode = HttpStatusCode.INTERNAL_SERVER_ERROR,
    message: String,
    cause: Throwable? = null
) : IOException(message) {

    init {
        cause?.let { initCause(it) }
    }

    companion object {

        const val ERROR_MESSAGE_HEADER = "x-wtw-errormessage"

        fun from(e: Exception): RestServiceException {
            val cause = e.cause as Exception?
            if (cause != null && cause is RestServiceException)
                return cause
            if (e is RestServiceException)
                return e
            return RestServiceException(HttpStatusCode.INTERNAL_SERVER_ERROR, e.message ?: "", cause = e)
        }

        fun from(response: Response): RestServiceException {
            val header = response.headers(ERROR_MESSAGE_HEADER).joinToString(separator = "\n")
            if (header.isNotBlank())
                return RestServiceException(HttpStatusCode.from(response.code), header)
            val body = response.body?.string()
                ?: return RestServiceException(HttpStatusCode.from(response.code), "Unknown error")
            if (body.startsWith("{")) {
                try {
                    val errorObject = GsonSingleton.getInstance().fromJson(body, RestServiceErrorObject::class.java)!!
                    return RestServiceException(HttpStatusCode.from(errorObject.code!!), errorObject.message!!)
                } catch (_: Exception) {
                }
            }
            return RestServiceException(HttpStatusCode.from(response.code), body)
        }

    }

    override fun toString(): String {
        var msg = "$statusCode - $message"
        cause?.let { msg += "\nCaused by: $it" }
        return msg
    }

}